services/gossip/adapter/tcp/outgoing_connection.go

```diff
@@ -52,20 +52,37 @@ func newOutgoingConnection(peer adapter.TransportPeer, parentLogger log.Logger,
 	queue.networkAddress = networkAddress
 	queue.Disable() // until connection is established
 
+	sendErrors, sendQueueErrors := generateMetrics(peerHexAddress, metricFactory)
+
 	client := &outgoingConnection{
 		logger:          logger,
 		sharedMetrics:   sharedMetrics,
 		metricRegistry:  metricFactory,
 		config:          transportConfig,
 		queue:           queue,
 		peerHexAddress:  peerHexAddress,
-		sendErrors:      metricFactory.NewGauge(fmt.Sprintf("Gossip.OutgoingConnection.SendError.%s.Count", peerHexAddress)),
-		sendQueueErrors: metricFactory.NewGauge(fmt.Sprintf("Gossip.OutgoingConnection.EnqueueErrors.%s.Count", peerHexAddress)),
+		sendErrors:      sendErrors,
+		sendQueueErrors: sendQueueErrors,
 	}
 
 	return client
 }
 
+// This is a round-about way to clean up these metrics that can be left over from previous connection
+func generateMetrics(peerHexAddress string, metricFactory metric.Registry) (*metric.Gauge, *metric.Gauge){
+	sendErrorsName := fmt.Sprintf("Gossip.OutgoingConnection.SendError.%s.Count", peerHexAddress)
+	sendErrorMetric := metricFactory.Get(sendErrorsName)
+	if sendErrorMetric != nil {
+		metricFactory.Remove(sendErrorMetric)
+	}
+	sendQueueErrorsName := fmt.Sprintf("Gossip.OutgoingConnection.EnqueueErrors.%s.Count", peerHexAddress)
+	sendQueueErrorMetric := metricFactory.Get(sendQueueErrorsName)
+	if sendQueueErrorMetric != nil {
+		metricFactory.Remove(sendQueueErrorMetric)
+	}
+	return metricFactory.NewGauge(sendErrorsName),  metricFactory.NewGauge(sendQueueErrorsName)
```

Comment:
this will work when there is only one goroutine trying to add the new metric. when two race it could lead to panic.

I see two alternatives:
1. add a reference count to each metric, then whenever you `NewGauge` for an existing metric you increase the reference count and return the preexisting metric. then, `remove` kills it only when the reference count 
2. `NewGauge` could always replace under lock in case the old name is already registered. 

each method (the two alternatives and your solution) has advantages and disadvantages. what do you think?

Reply:
All the connections are open and closed under lock for the outgoing_connections so it is not needed.
the newgauge doesn't need to know there is a "replace" thing its not its job.